src/Oracle/Protocol/OracleHttpProtocol.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using JArray = Newtonsoft.Json.Linq.JArray;
+using JObject = Newtonsoft.Json.Linq.JObject;
+
+namespace Neo.Plugins
+{
+    public class OracleHttpProtocol : IOracleProtocol
+    {
+        public int Timeout { get; set; } = 5000;
+        public bool AllowPrivateHost { get; set; } = false;
+        public readonly string[] AllowedFormats = new string[] { "application/json" };
+
+        public byte[] Request(ulong requestId, string url, string filter)
+        {
+            Utility.Log(nameof(OracleHttpProtocol), LogLevel.Debug, $"Request: {url}");
+
+            Uri.TryCreate(url, UriKind.Absolute, out var uri);
+            if (!AllowPrivateHost && IsInternal(Dns.GetHostEntry(uri.Host)))
+            {
+                // Don't allow private host in order to prevent SSRF
+
+                throw new InvalidOperationException("PolicyError");
+            }
+            using var handler = new HttpClientHandler
+            {
+                // TODO: Accept all certificates
+                ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+            };
+            using var client = new HttpClient(handler);
+            client.DefaultRequestHeaders.Add("Accept", string.Join(",", AllowedFormats));
+
+            Task<HttpResponseMessage> result = client.GetAsync(uri);
+
+            if (!result.Wait(Timeout)) throw new InvalidOperationException("Timeout");
+            if (!result.Result.IsSuccessStatusCode) throw new InvalidOperationException("Response error");
+            if (!AllowedFormats.Contains(result.Result.Content.Headers.ContentType.MediaType)) throw new InvalidOperationException("ContentType it's not allowed");
+
+            var taskRet = result.Result.Content.ReadAsStringAsync();
+            if (!taskRet.Wait(Timeout)) throw new InvalidOperationException("Timeout");
```

Comment:
We should subtract the time that was spent in line 38?

Reply:
Done, Replaced it by `Timeout/2`.